File: emarket/app/channels/chat_room_channel.rb
class ChatRoomChannel < ApplicationCable::Channel
  def subscribed
    @chat_room = ChatRoom.find(params[:chat_id])
    stream_for @chat_room
  end

  def speak(data)
    messaggio = @chat_room.messaggi.create!(
      contenuto: data['message'],
      mittente: current_user
    )
    message_html = ApplicationController.render(
      partial: 'messaggi/messaggio',
      locals: { messaggio: messaggio }
    )
    ChatRoomChannel.broadcast_to(@chat_room, { message_html: message_html })
    end

  private

  def render_message(messaggio)
    ApplicationController.render(partial: 'messaggi/messaggio', locals: { messaggio: messaggio })
  end
end

File: emarket/app/javascript/channels/chat_room_channel.js
import consumer from "./consumer";

document.addEventListener("turbolinks:load", () => {
  const chatRoomElement = document.getElementById("chat-room");
  if (!chatRoomElement) return;

  const chatRoomId = chatRoomElement.dataset.chatRoomId;

  const subscription = consumer.subscriptions.create(
    { channel: "ChatRoomChannel", chat_id: chatRoomId },
    {
      received(data) {
        // Aggiorna solo il contenitore della chat
        const messagesContainer = document.getElementById("messages");
        
        // Aggiungi il messaggio ricevuto
        messagesContainer.insertAdjacentHTML("beforeend", data.message_html);
        
        // Scorri in basso automaticamente
        messagesContainer.scrollTop = messagesContainer.scrollHeight;
      },
    }
  );

  // Aggiungi la possibilità di inviare messaggi tramite il pulsante 
  const messageInput = document.getElementById("message_input");
  const sendButton = document.getElementById("send_message_button");

  const sendMessage = () => {
    const message = messageInput.value.trim();
    if (message === "") return;

    subscription.perform("speak", { message });
    messageInput.value = ""; // Resetta il campo di input
  };

  sendButton.addEventListener("click", sendMessage);
  messageInput.addEventListener("keypress", (e) => {
    if (e.key === "Enter") {
      e.preventDefault();
      sendMessage();
    }
  });
});
